rasa/core/policies/memoization.py

```diff
@@ -246,31 +244,41 @@ def load(cls, path: Text) -> "MemoizationPolicy":
 
 class AugmentedMemoizationPolicy(MemoizationPolicy):
     """The policy that remembers examples from training stories
-        for `max_history` turns.
-
-        If it is needed to recall turns from training dialogues
-        where some slots might not be set during prediction time,
-        add relevant stories without such slots to training data.
-        E.g. reminder stories.
-
-        Since `slots` that are set some time in the past are
-        preserved in all future feature vectors until they are set
-        to None, this policy has a capability to recall the turns
-        up to `max_history` from training stories during prediction
-        even if additional slots were filled in the past
-        for current dialogue.
+    for `max_history` turns.
+
+    If it is needed to recall turns from training dialogues
+    where some slots might not be set during prediction time,
+    add relevant stories without such slots to training data.
+    E.g. reminder stories.
+
+    Since `slots` that are set some time in the past are
+    preserved in all future feature vectors until they are set
+    to None, this policy has a capability to recall the turns
+    up to `max_history` from training stories during prediction
+    even if additional slots were filled in the past
+    for current dialogue.
     """
 
     @staticmethod
-    def _back_to_the_future_again(tracker) -> Optional[DialogueStateTracker]:
+    def _back_to_the_future(
+        tracker: DialogueStateTracker,
+        again: bool = False,
+        max_history: Optional[int] = None,
+    ) -> Optional[DialogueStateTracker]:
         """Send Marty to the past to get
-            the new featurization for the future"""
+        the new featurization for the future"""
 
         idx_of_first_action = None
         idx_of_second_action = None
 
+        applied_events = tracker.applied_events()
+        # We don't need to look at events that occured before `max_history`
+        start_index = (
```

Comment:
why don't we do this once in `recall` instead of here where this is part of the loop?